### PR TITLE
Interior/fuel selector (fuel system)

### LINKED ISSUE #52: Fuel system
Create the fuel system for the C182S according POH as it works completly different to older Cessna single props.
1.)
It should create affect fuel consumption when engaged, depending by mixture and also able to flood the engine.

2.)it should also be affectd by the engine and outside temperature.

### PR DESCRIPTION
**Pull request by hbeni**
Hi there,
i filed this PR (mission is to fix #52 and enable #107) mainly to track open todos and streamline the discussions from #52 and  3f5357b206:

 - [x] rate of fuel transfer of tanks (crossfeed), especially honoring of gravitiy forces and relative positions
 - [x] there seems to be a small leak when transferring fuel from one tank to another
 - [x] failure of fuel pipe
 - [x] leakings of tanks and sumps
 - [x] make autopilot handle fuel transfer correctly
 - [ ] ~~making the failure states toggable in the failure manager~~ impossible for now
 - [x] hiding the internal tanks (they are still visible in fuel dialog for easier debug)
 - [x] correct drain and tank locations
 - [x] add drain options to menu
 - [x] correct fuel flow of system 
 - [x] add wing drain action points to model

### CODE REVIEW

**Comment by HHS81** [comment]
I see some error messages: 
```
Failed to tie property fcs/ to object methods

Failed to tie property fcs/ to object methods

Failed to tie property fcs/ to object methods

Failed to tie property fcs/ to object methods

Failed to tie property fcs/ to object methods 
```
 Maybe [this](https://forum.flightgear.org/viewtopic.php?f=4&t=28820) gives a hint.
And even if the wing tanks are empty, the engine won't stop, as it gets magically unlimited fuel supply from fuel pipe, fuel sumps...

Otherwise really great!

Should I merge it anyway?

**Comment by hbeni** [comment]
Please don't merge for now, as i try to get the infrastructure fully done.

**Comment by HHS81** [comment]
>  I didn't know about that bug and will revert the explicit density value given (despite the jsbsim manual says, giving AVGAS as type will override density value!).

Hmm, that's interesting. 
The last time I tested it, it ignored the Fuel type. Now it uses it 😲 

So with 88 Gallons fuel we get 529.4 lbs fuel weight.
According the [POH](http://tssflyingclub.org/documents/C182S_POH.pdf) p.6-12 it should be 528. And according to thie page it uses a density of 6.0 lbs/gallon.

**Comment by hbeni** [comment]
> Failed to tie property fcs/ to object methods
I have noticed them too but cannot track them down. using "grep" i dod not find any occurence that catched my eye...

`beni@segin:~/Dokumente/FGFS/work/c182s$ grep "fcs/" -r *
c182s.xml:          <function name="fcs/nose-gear-left-brake-steer-factor">
c182s.xml:               <property>fcs/left-brake-cmd-norm</property>
c182s.xml:          <function name="fcs/nose-gear-right-brake-steer-factor">
c182s.xml:               <property>fcs/right-brake-cmd-norm</property>
c182s.xml:                <input>fcs/nose-gear-right-brake-steer-factor</input>
c182s.xml:                <input>fcs/nose-gear-left-brake-steer-factor</input>
c182s.xml:          <function name="fcs/brake-steer-influence">
c182s.xml:                   <independentVar lookup="row">fcs/brake-sum</independentVar>
c182s.xml:                   <independentVar lookup="column">fcs/steer-cmd-norm</independentVar>
c182s.xml:          <output>fcs/steer-pos-deg</output> 
c182s.xml:                <input>fcs/elevator-cmd-norm</input>
c182s.xml:                <input>fcs/pitch-trim-cmd-norm</input>
c182s.xml:                <input>fcs/pitch-trim-sum</input>
c182s.xml:                <output>fcs/elevator-pos-rad</output>
c182s.xml:                <input>fcs/elevator-pos-deg</input>
c182s.xml:                <output>fcs/elevator-pos-norm</output>
c182s.xml:      <input>fcs/aileron-cmd-norm</input>
c182s.xml:      <input>fcs/roll-trim-cmd-norm</input>
c182s.xml:      <output>fcs/roll-trim-sum</output>
c182s.xml:      <input>fcs/roll-trim-sum</input>
c182s.xml:      <output>fcs/left-aileron-pos-norm</output>
c182s.xml:      <input>fcs/roll-trim-sum</input>
c182s.xml:      <output>fcs/left-aileron-pos-rad</output>
c182s.xml:      <input>fcs/roll-trim-sum</input>
c182s.xml:      <output>fcs/right-aileron-pos-norm</output>
c182s.xml:      <input>fcs/roll-trim-sum</input>
c182s.xml:      <output>fcs/right-aileron-pos-rad</output>
c182s.xml:                <input>fcs/rudder-cmd-norm</input>
c182s.xml:                <input>fcs/yaw-trim-cmd-norm</input>
c182s.xml:                <input>fcs/yaw-trim-sum</input>
c182s.xml:                <output>fcs/rudder-pos-rad</output>
c182s.xml:                <input>fcs/rudder-pos-deg</input>
c182s.xml:                <output>fcs/rudder-pos-norm</output>
c182s.xml:                <input>fcs/flap-cmd-norm</input>
c182s.xml:                <output>fcs/flap-pos-norm</output>
c182s.xml:              <input>fcs/flap-pos-norm</input>
c182s.xml:              <output>fcs/flap-pos-deg</output>
c182s.xml:                        <independentVar>fcs/flap-pos-deg</independentVar>
c182s.xml:                    <property>fcs/elevator-pos-rad</property>
c182s.xml:                        <independentVar>fcs/flap-pos-deg</independentVar>
c182s.xml:                          <independentVar lookup="column">fcs/flap-pos-deg</independentVar>
c182s.xml:                    <property>fcs/mag-elevator-pos-rad</property>
c182s.xml:                    <property>fcs/rudder-pos-rad</property>
c182s.xml:          <property>fcs/left-aileron-pos-rad</property>
c182s.xml:                    <property>fcs/rudder-pos-rad</property>
c182s.xml:                    <property>fcs/elevator-pos-rad</property>
c182s.xml:                        <independentVar>fcs/flap-pos-deg</independentVar>
c182s.xml:                    <property>fcs/left-aileron-pos-rad</property>
c182s.xml:                    <property>fcs/rudder-pos-rad</property>
c182t.xml:          <function name="fcs/nose-gear-left-brake-steer-factor">
c182t.xml:               <property>fcs/left-brake-cmd-norm</property>
c182t.xml:          <function name="fcs/nose-gear-right-brake-steer-factor">
c182t.xml:               <property>fcs/right-brake-cmd-norm</property>
c182t.xml:                <input>fcs/nose-gear-right-brake-steer-factor</input>
c182t.xml:                <input>fcs/nose-gear-left-brake-steer-factor</input>
c182t.xml:          <function name="fcs/brake-steer-influence">
c182t.xml:                   <independentVar lookup="row">fcs/brake-sum</independentVar>
c182t.xml:                   <independentVar lookup="column">fcs/steer-cmd-norm</independentVar>
c182t.xml:          <output>fcs/steer-pos-deg</output> 
c182t.xml:                <input>fcs/elevator-cmd-norm</input>
c182t.xml:                <input>fcs/pitch-trim-cmd-norm</input>
c182t.xml:                <input>fcs/pitch-trim-sum</input>
c182t.xml:                <output>fcs/elevator-pos-rad</output>
c182t.xml:                <input>fcs/elevator-pos-deg</input>
c182t.xml:                <output>fcs/elevator-pos-norm</output>
c182t.xml:                <input>fcs/aileron-cmd-norm</input>
c182t.xml:                <input>fcs/roll-trim-cmd-norm</input>
c182t.xml:                <input>fcs/roll-trim-sum</input>
c182t.xml:                <output>fcs/left-aileron-pos-rad</output>
c182t.xml:                <input>fcs/left-aileron-pos-deg</input>
c182t.xml:                <output>fcs/left-aileron-pos-norm</output>
c182t.xml:                <input>fcs/roll-trim-sum</input>
c182t.xml:                <output>fcs/right-aileron-pos-rad</output>
c182t.xml:                <input>fcs/right-aileron-pos-deg</input>
c182t.xml:                <output>fcs/right-aileron-pos-norm</output>
c182t.xml:                <input>fcs/rudder-cmd-norm</input>
c182t.xml:                <input>fcs/yaw-trim-cmd-norm</input>
c182t.xml:                <input>fcs/yaw-trim-sum</input>
c182t.xml:                <output>fcs/rudder-pos-rad</output>
c182t.xml:                <input>fcs/rudder-pos-deg</input>
c182t.xml:                <output>fcs/rudder-pos-norm</output>
c182t.xml:                <input>fcs/flap-cmd-norm</input>
c182t.xml:                <output>fcs/flap-pos-norm</output>
c182t.xml:              <input>fcs/flap-pos-norm</input>
c182t.xml:              <output>fcs/flap-pos-deg</output>
c182t.xml:                        <independentVar>fcs/flap-pos-deg</independentVar>
c182t.xml:                    <property>fcs/elevator-pos-rad</property>
c182t.xml:                        <independentVar>fcs/flap-pos-deg</independentVar>
c182t.xml:                          <independentVar lookup="column">fcs/flap-pos-deg</independentVar>
c182t.xml:                    <property>fcs/mag-elevator-pos-rad</property>
c182t.xml:                    <property>fcs/rudder-pos-rad</property>
c182t.xml:          <property>fcs/left-aileron-pos-rad</property>
c182t.xml:                    <property>fcs/rudder-pos-rad</property>
c182t.xml:                    <property>fcs/elevator-pos-rad</property>
c182t.xml:                        <independentVar>fcs/flap-pos-deg</independentVar>
c182t.xml:                    <property>fcs/left-aileron-pos-rad</property>
c182t.xml:                    <property>fcs/rudder-pos-rad</property>
Models/c182s.xml:   <property>fdm/jsbsim/fcs/steer-pos-deg</property>
Systems/indicated-airspeed.xml:                    <independentVar lookup="column">fcs/flap-pos-deg</independentVar>
Systems/c182s-ground-effects.xml:        <summer name="fcs/tiedown-left-mag-error">
Systems/c182s-ground-effects.xml:        <summer name="fcs/tiedown-right-mag-error">
Systems/c182s-ground-effects.xml:        <summer name="fcs/tiedown-tail-mag-error">
Systems/c182s-ground-effects.xml:        <pid name="fcs/tiedown-left-magnitude-pid">
Systems/c182s-ground-effects.xml:            <input>fcs/tiedown-left-mag-error</input>
Systems/c182s-ground-effects.xml:        <pid name="fcs/tiedown-right-magnitude-pid">
Systems/c182s-ground-effects.xml:            <input>fcs/tiedown-right-mag-error</input>
Systems/c182s-ground-effects.xml:        <pid name="fcs/tiedown-tail-magnitude-pid">
Systems/c182s-ground-effects.xml:            <input>fcs/tiedown-tail-mag-error</input>
Systems/c182s-ground-effects.xml:        <switch name="fcs/tiedown-left-magnitude">
Systems/c182s-ground-effects.xml:            <test logic="AND" value="fcs/tiedown-left-magnitude-pid">
Systems/c182s-ground-effects.xml:        <switch name="fcs/tiedown-right-magnitude">
Systems/c182s-ground-effects.xml:            <test logic="AND" value="fcs/tiedown-right-magnitude-pid">
Systems/c182s-ground-effects.xml:        <switch name="fcs/tiedown-tail-magnitude">
Systems/c182s-ground-effects.xml:            <test logic="AND" value="fcs/tiedown-tail-magnitude-pid">
Systems/c182s-stallhorn.xml:                                <independentVar>fcs/flap-pos-deg</independentVar>
Systems/c182s-stallhorn.xml:                            <independentVar lookup="column">fcs/flap-pos-deg</independentVar>
Systems/c182s-sounds.xml:                                <independentVar>fcs/flap-pos-deg</independentVar>
Systems/c182s-sounds.xml:                            <independentVar lookup="column">fcs/flap-pos-deg</independentVar>`

**Comment by HHS81** [comment]
Hi,
I see the fuel consume is higher than the actual fuel flow by the engine. 
The engine model in JSBSim is quite sophisticated, so it actually simulates the fuel consume of the engine very well - it is represented in the properties as  `fdm/jsbsim/propulsion/fuel-flow-rate-gph` or `fdm/jsbsim/propulsion/fuel-flow-rate-pps`
I would expect, that when the engine is running, aux pump is off, and aircraft is level, the fuel flow between all tanks and engine manifold is equal the engine fuel flow. So I would expect that the `/fdm/jsbsim/propulsion/tank[3]/external-flow-rate-pps` is under normal circumstances the same as `fdm/jsbsim/propulsion/fuel-flow-rate-pps`.

**Comment by hbeni** [comment]
Hi, im already on it

Am 2017-09-10 11:47, schrieb HHS81:

**Comment by hbeni** [comment]
I just pushed some commits; the engines thirst should be gone now.
I did some tests under full throttle and fuel flow is much slower now (but i don't know if its correct now! please test; but be aware that i don't know how to fix this further).


@HHS81 @gilbertohasnofb For the last big point (in my opinion), the tank drain and transfer rates, i need some real life information:
> rate of fuel transfer of tanks (crossfeed), especially honoring of gravitiy forces and relative positions
1. How much fuel transfers per second from left wing to right wing, if BOTH selected and uneven level?
2. How fast do the tanks drain in reality?

**Comment by HHS81** [comment]
> I did some tests under full throttle and fuel flow is much slower now 

That means, it isn't equal the engine fuel flow?

> How much fuel transfers per second from left wing to right wing, if BOTH selected and uneven level?
No, idea, depends probably on the angle

> How fast do the tanks drain in reality?
No idea

**Comment by hbeni** [comment]
ok, then i marked the todo in the list as checked since i have no clue either. This can be finetuned later anyway; the core feature is here. Under normal flying circumstances disbalance shouldn't happen anyway but the user can force/fix it now with the fuel selector :)
Also for the drain rates, i choosed values that are clearly visible but dont drain too fast (so we have something to watch :)

**Comment by hbeni** [comment]
So, now i will focus on adding action points to the wings and adjust the internal tanks postions a little (the wing tanks are already in the correct xyz coords, i guess?)

**Comment by HHS81** [comment]
The last time I checked it was in the correct xyz-coords, according POH

**Comment by HHS81** [comment]
@hbeni  fuel flow is still too high....

**Comment by hbeni** [comment]
An idea came to my mind tonight, i hope it solves the issue. I will try!

File: Systems/fuel.xml [hunk 1]
```diff
@@ -368,7 +368,7 @@ Implementation notice:
             <input>fuel/pump-flow-rate-pps</input>
             <input>-/systems/fuel/tank[5]/leak-flow-rate-pps</input>
             <input>-fuel/manifold-dissipate-rate-pps</input>
-            <input>-/fdm/jsbsim/propulsion/engine/fuel-flow-rate-pps</input>
+            <!--    <input>-/fdm/jsbsim/propulsion/engine/fuel-flow-rate-pps</input>   NOT NECESSARY: engine sucks already (see feed command) -->
```
Comment (hbeni) [inline Systems/fuel.xml]: Tested by using up a complete tank at various power settings at 32x simulation rate.
when the engine stopped, it showed ~530 lbs used, equalling my initial filling in the system.
=> There is no leak anymore :)

**Comment by hbeni** [comment]
@HHS81 @gilbertohasnofb 
i consider the basic fuel system done by now.
Please test :)

If it is OK for you, i will start another branch for the complex engine procedures from the then-merged master. This will take some time, tough.

**Comment by HHS81** [comment]
I couldn't test it yet,  but Sounds all o.k. From the procedure is not much missing - see my comments from the complex startup issue. 

Many thanks Benedikt :thumbs up:

**Comment by hbeni** [comment]
> Many thanks Benedikt :thumbs up:
this was fun :)

**Comment by HHS81** [comment]
Could solve the issue with my laptop - faster than I thought.

**Comment by HHS81** [comment]
O.k.  fuel flow is o.k. now
But the position of the wing drain is wrong:
fgfs-screen-001
That's actually the exterior light, which is engaged together with rear interior light, as you can see here on the image.
It is difficult to tell where it is, but according some images on the web it could be closer to the cabin, near the flaps

**Comment by hbeni** [comment]
Okay, i will research further and correct this. Please stand by!

**Comment by hbeni** [comment]
@HHS81 corrected this according to some internet pics.

For the priming stuff i already started a private local branch which i will push soon

**Comment by hbeni** [comment]
work continued in #107

**Pull request closed without merging**